<a href="https://colab.research.google.com/github/Anurag-Raj-Shukla/Chess-Engine/blob/main/Next_Word_N_gram.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# NLP Trigram next word/phrase suggestor


# Task
Build a trigram model using the wikipedia library to predict the next word/phrase.

## Install wikipedia library

### Subtask:
Install the necessary library using pip.


**Reasoning**:
The subtask is to install the `wikipedia` library. I will use the `pip` command with the `install` option to accomplish this.



In [ ]:
%pip install wikipedia

## Load and preprocess data

### Subtask:
Load the text data using the wikipedia library and preprocess it by converting to lowercase, removing punctuation, and tokenizing into words.


**Reasoning**:
Load the text data using the wikipedia library, convert it to lowercase, remove punctuation, and tokenize it into words.



In [ ]:
import wikipedia
import string

# Search for a Wikipedia page and get its content
page_title = "Artificial intelligence"
try:
    page_content = wikipedia.page(page_title).content
except wikipedia.exceptions.PageError:
    print(f"Page '{page_title}' not found.")
    page_content = ""
except wikipedia.exceptions.DisambiguationError as e:
    print(f"Disambiguation page for '{page_title}'. Options: {e.options}")
    page_content = wikipedia.page(e.options[0]).content


# Convert to lowercase
lowercase_text = page_content.lower()

# Remove punctuation
translator = str.maketrans('', '', string.punctuation)
cleaned_text = lowercase_text.translate(translator)

# Tokenize into words
words = cleaned_text.split()

print(f"First 20 tokens: {words[:20]}")

First 20 tokens: ['artificial', 'intelligence', 'ai', 'is', 'the', 'capability', 'of', 'computational', 'systems', 'to', 'perform', 'tasks', 'typically', 'associated', 'with', 'human', 'intelligence', 'such', 'as', 'learning']


## Create trigrams

### Subtask:
Generate trigrams (sequences of three consecutive words) from the tokenized text.


**Reasoning**:
Generate trigrams from the tokenized words list and print the first 10 trigrams.



In [ ]:
trigrams = []
for i in range(len(words) - 2):
    trigram = (words[i], words[i+1], words[i+2])
    trigrams.append(trigram)

print("First 10 trigrams:")
for trigram in trigrams[:10]:
    print(trigram)

First 10 trigrams:
('artificial', 'intelligence', 'ai')
('intelligence', 'ai', 'is')
('ai', 'is', 'the')
('is', 'the', 'capability')
('the', 'capability', 'of')
('capability', 'of', 'computational')
('of', 'computational', 'systems')
('computational', 'systems', 'to')
('systems', 'to', 'perform')
('to', 'perform', 'tasks')


## Build trigram frequency table

### Subtask:
Build a trigram frequency table to count how many times each trigram appears.


**Reasoning**:
Import Counter and build the trigram frequency table using the list of trigrams, then print the 10 most common trigrams.



In [ ]:
from collections import Counter

trigram_frequencies = Counter(trigrams)

print("10 most common trigrams:")
for trigram, count in trigram_frequencies.most_common(10):
    print(f"{trigram}: {count}")

10 most common trigrams:
('of', 'artificial', 'intelligence'): 11
('the', 'ability', 'to'): 7
('artificial', 'general', 'intelligence'): 6
('as', 'well', 'as'): 6
('can', 'be', 'used'): 6
('the', 'study', 'of'): 6
('of', 'ai', 'research'): 5
('the', 'use', 'of'): 5
('be', 'able', 'to'): 5
('a', 'set', 'of'): 5


## Build next word prediction function

### Subtask:
Create a function that takes two words as input and predicts the next word based on the trigram frequencies.


**Reasoning**:
Define the function to predict the next word based on trigram frequencies.



In [ ]:
def predict_next_word(word1, word2, trigram_frequencies):
    """
    Predicts the next word based on trigram frequencies.

    Args:
        word1: The first word of the bigram.
        word2: The second word of the bigram.
        trigram_frequencies: A Counter object containing trigram frequencies.

    Returns:
        The most frequent third word that follows the bigram, or None if no
        matching trigrams are found.
    """
    matching_trigrams = {}
    for trigram, frequency in trigram_frequencies.items():
        if trigram[0] == word1 and trigram[1] == word2:
            matching_trigrams[trigram[2]] = frequency

    if matching_trigrams:
        # Find the word with the highest frequency
        predicted_word = max(matching_trigrams, key=matching_trigrams.get)
        return predicted_word
    else:
        return None

# Example usage:
word1 = "artificial"
word2 = "intelligence"
prediction = predict_next_word(word1, word2, trigram_frequencies)
print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

word1 = "this"
word2 = "is"
prediction = predict_next_word(word1, word2, trigram_frequencies)
print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

word1 = "nonexistent"
word2 = "bigram"
prediction = predict_next_word(word1, word2, trigram_frequencies)
print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

Given 'artificial intelligence', the predicted next word is: is
Given 'this is', the predicted next word is: the
Given 'nonexistent bigram', the predicted next word is: None


## Test the model

### Subtask:
Test the prediction function with some example word pairs.


**Reasoning**:
Call the predict_next_word function with different word pairs and print the results to test its functionality.



In [ ]:
# Test the prediction function with some example word pairs
test_cases = [
    ("artificial", "intelligence"),
    ("machine", "learning"),
    ("data", "science"),
    ("this", "is"),
    ("a", "set"),
    ("apple", "banana") # Example of words not expected to be found
]

for word1, word2 in test_cases:
    prediction = predict_next_word(word1, word2, trigram_frequencies)
    print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

Given 'artificial intelligence', the predicted next word is: is
Given 'machine learning', the predicted next word is: models
Given 'data science', the predicted next word is: None
Given 'this is', the predicted next word is: the
Given 'a set', the predicted next word is: of
Given 'apple banana', the predicted next word is: None


## Summary:

### Data Analysis Key Findings

*   The `wikipedia` library was successfully installed and used to load text data from the "Artificial intelligence" page.
*   The loaded text was successfully preprocessed by converting it to lowercase, removing punctuation, and tokenizing it into words.
*   Trigrams (sequences of three consecutive words) were successfully generated from the tokenized text.
*   A trigram frequency table was built using `collections.Counter`, effectively counting the occurrences of each unique trigram.
*   A function `predict_next_word` was created which takes two words as input and predicts the next word based on the trigram frequencies, returning the most frequent third word or `None` if no matching trigrams are found.
*   Testing of the prediction function showed that it successfully predicted the next word for bigrams present in the training data (e.g., "artificial intelligence" predicted "is") and returned `None` for bigrams not found in the data (e.g., "apple banana").

### Insights or Next Steps

*   The current model uses simple frequency counts. Exploring smoothing techniques (like Add-One smoothing) could improve predictions for less frequent bigrams.
*   The model's performance is limited by the size and specificity of the single Wikipedia page used for training. Training on a larger, more diverse corpus would likely yield better prediction results.


# Task
Build a trigram language model with Add-One smoothing using the `wikipedia` library and train it on a diverse corpus to predict the next word or phrase.

## Load and preprocess data

### Subtask:
Load text data from a more diverse corpus (e.g., multiple Wikipedia pages, or other text sources) and preprocess it by converting to lowercase, removing punctuation, and tokenizing into words.


**Reasoning**:
Load text data from multiple Wikipedia pages, preprocess it by converting to lowercase, removing punctuation, and tokenizing into words as per the instructions.



In [ ]:
import wikipedia
import string

# Create a list of Wikipedia page titles for a diverse corpus
page_titles = ["History", "Science", "Art", "Technology", "Literature", "Philosophy"]
combined_content = ""

# Iterate through the list of page titles and get content
for page_title in page_titles:
    print(f"Processing page: {page_title}")
    try:
        page_content = wikipedia.page(page_title).content
        combined_content += page_content
    except wikipedia.exceptions.PageError:
        print(f"Page '{page_title}' not found.")
    except wikipedia.exceptions.DisambiguationError as e:
        print(f"Disambiguation page for '{page_title}'. Options: {e.options}")
        try:
            page_content = wikipedia.page(e.options[0]).content
            combined_content += page_content
        except wikipedia.exceptions.PageError:
            print(f"Could not get content from the first option for '{page_title}'.")

# Convert to lowercase
lowercase_text = combined_content.lower()

# Remove punctuation
translator = str.maketrans('', '', string.punctuation)
cleaned_text = lowercase_text.translate(translator)

# Tokenize into words
words = cleaned_text.split()

print(f"\nFirst 20 tokens from the diverse corpus: {words[:20]}")

Processing page: History
Processing page: Science
Disambiguation page for 'Science'. Options: ['Sincé', 'Sińce', 'Since (film)', 'Since (album)', 'Since (rapper)', 'Snice']
Processing page: Art
Processing page: Technology
Processing page: Literature
Processing page: Philosophy

First 20 tokens from the diverse corpus: ['history', 'is', 'the', 'systematic', 'study', 'of', 'the', 'past', 'focusing', 'primarily', 'on', 'the', 'human', 'past', 'as', 'an', 'academic', 'discipline', 'it', 'analyses']


## Create trigrams

### Subtask:
Generate trigrams (sequences of three consecutive words) from the tokenized text.


**Reasoning**:
Generate trigrams from the tokenized words list and print the first 10 trigrams.



In [ ]:
trigrams = []
for i in range(len(words) - 2):
    trigram = (words[i], words[i+1], words[i+2])
    trigrams.append(trigram)

print("First 10 trigrams:")
for trigram in trigrams[:10]:
    print(trigram)

First 10 trigrams:
('history', 'is', 'the')
('is', 'the', 'systematic')
('the', 'systematic', 'study')
('systematic', 'study', 'of')
('study', 'of', 'the')
('of', 'the', 'past')
('the', 'past', 'focusing')
('past', 'focusing', 'primarily')
('focusing', 'primarily', 'on')
('primarily', 'on', 'the')


## Build trigram frequency table

### Subtask:
Create a frequency distribution of the trigrams to count how many times each trigram appears.


**Reasoning**:
Import Counter and build the trigram frequency table using the list of trigrams, then print the 10 most common trigrams.



In [ ]:
from collections import Counter

trigram_frequencies = Counter(trigrams)

print("10 most common trigrams:")
for trigram, count in trigram_frequencies.most_common(10):
    print(f"{trigram}: {count}")

10 most common trigrams:
('the', 'study', 'of'): 21
('the', 'history', 'of'): 18
('as', 'well', 'as'): 18
('such', 'as', 'the'): 17
('the', 'philosophy', 'of'): 16
('the', 'invention', 'of'): 15
('in', 'the', 'form'): 12
('the', 'form', 'of'): 12
('one', 'of', 'the'): 12
('the', 'emergence', 'of'): 12


## Build next word prediction function

### Subtask:
Create a function that takes two words as input and predicts the next word based on the trigram frequencies, incorporating Add-One smoothing.


**Reasoning**:
Define the function to predict the next word based on trigram frequencies with Add-One smoothing, calculate the vocabulary size, and test the function with example word pairs.



In [ ]:
def predict_next_word_smoothed(word1, word2, trigram_frequencies, vocabulary_size):
    """
    Predicts the next word based on trigram frequencies with Add-One smoothing.

    Args:
        word1: The first word of the bigram.
        word2: The second word of the bigram.
        trigram_frequencies: A Counter object containing trigram frequencies.
        vocabulary_size: The total number of unique words in the corpus.

    Returns:
        The most frequent third word that follows the bigram after smoothing,
        or None if no matching trigrams are found and the vocabulary is empty.
    """
    matching_trigrams = {}
    for trigram, frequency in trigram_frequencies.items():
        if trigram[0] == word1 and trigram[1] == word2:
            matching_trigrams[trigram[2]] = frequency

    # Apply Add-One smoothing
    smoothed_frequencies = {}
    unique_words = set(words) # Get the vocabulary from the global 'words' list

    for word in unique_words:
        smoothed_frequencies[word] = matching_trigrams.get(word, 0) + 1

    if smoothed_frequencies:
        # Find the word with the highest smoothed frequency
        predicted_word = max(smoothed_frequencies, key=smoothed_frequencies.get)
        return predicted_word
    else:
        return None

# Calculate the vocabulary size
vocabulary_size = len(set(words))
print(f"Vocabulary size: {vocabulary_size}")

# Test the predict_next_word_smoothed function with some example word pairs
test_cases = [
    ("the", "study"),
    ("history", "is"),
    ("as", "well"),
    ("apple", "banana") # Example of words not expected to be found
]

for word1, word2 in test_cases:
    prediction = predict_next_word_smoothed(word1, word2, trigram_frequencies, vocabulary_size)
    print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

Vocabulary size: 4977
Given 'the study', the predicted next word is: of
Given 'history is', the predicted next word is: a
Given 'as well', the predicted next word is: as
Given 'apple banana', the predicted next word is: coloured


## Test the model

### Subtask:
Test the prediction function with some example word pairs.


**Reasoning**:
Call the predict_next_word_smoothed function with different word pairs and print the results to test its functionality.



In [ ]:
# Test the prediction function with some example word pairs
test_cases = [
    ("the", "study"),
    ("history", "is"),
    ("as", "well"),
    ("art", "history"),
    ("technology", "and"),
    ("nonexistent", "bigram") # Example of words not expected to be found
]

for word1, word2 in test_cases:
    prediction = predict_next_word_smoothed(word1, word2, trigram_frequencies, vocabulary_size)
    print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

Given 'the study', the predicted next word is: of
Given 'history is', the predicted next word is: a
Given 'as well', the predicted next word is: as
Given 'art history', the predicted next word is: another
Given 'technology and', the predicted next word is: nature
Given 'nonexistent bigram', the predicted next word is: coloured


## Evaluate and refine

### Subtask:
Evaluate the model's performance on the diverse corpus and explore ways to further refine the model.


**Reasoning**:
I need to evaluate the model's performance, identify limitations, and explore refinements. I will evaluate by calculating perplexity on a held-out test set and qualitatively assessing predictions. This will help identify limitations, such as the impact of smoothing and corpus dependence. Finally, I will discuss potential refinements based on these findings. Since I don't have a separate test set, I will simulate one by taking a small portion of the existing data.



In [ ]:
import numpy as np

# Split data for evaluation (simulate a test set)
# We'll use the last 100 bigrams for testing
if len(words) > 200: # Ensure enough data for a test set
    test_words = words[-200:]
    training_words = words[:-200]
else:
    test_words = words
    training_words = words

# Rebuild trigrams and frequencies on the training data
training_trigrams = []
for i in range(len(training_words) - 2):
    trigram = (training_words[i], training_words[i+1], training_words[i+2])
    training_trigrams.append(trigram)

training_trigram_frequencies = Counter(training_trigrams)
training_vocabulary_size = len(set(training_words))

# Calculate perplexity on the simulated test set
# Perplexity = (1/N * sum(-log2(P(w_i | w_{i-2}, w_{i-1}))))
# where N is the number of trigrams in the test set
# and P is the smoothed probability from the model.

log_prob_sum = 0
n_test_trigrams = 0

for i in range(len(test_words) - 2):
    w1, w2, w3 = test_words[i], test_words[i+1], test_words[i+2]

    # Calculate smoothed probability P(w3 | w1, w2)
    # P(w3 | w1, w2) = (count(w1, w2, w3) + 1) / (count(w1, w2) + Vocabulary Size)
    bigram_count = training_trigram_frequencies[(w1, w2, 'any_word')] -1 # Placeholder to count bigram occurrences
    # Need to count occurrences of the bigram (w1, w2)
    bigram_freq = sum(1 for trigram in training_trigrams if trigram[0] == w1 and trigram[1] == w2)

    smoothed_prob = (training_trigram_frequencies[(w1, w2, w3)] + 1) / (bigram_freq + training_vocabulary_size)

    if smoothed_prob > 0:
        log_prob_sum += np.log2(smoothed_prob)
        n_test_trigrams += 1
    else:
        # This case should theoretically not happen with Add-One smoothing
        # but including for robustness.
        print(f"Warning: Zero probability for trigram: ({w1}, {w2}, {w3})")


if n_test_trigrams > 0:
    average_log_prob = log_prob_sum / n_test_trigrams
    perplexity = 2 ** (-average_log_prob)
    print(f"\nPerplexity on simulated test set: {perplexity}")
else:
    print("\nNot enough test trigrams to calculate perplexity.")

# Qualitative assessment of predictions on the test set's bigrams
print("\nQualitative assessment of predictions on test set bigrams:")
for i in range(len(test_words) - 2):
    w1, w2, w3_actual = test_words[i], test_words[i+1], test_words[i+2]
    predicted_w3 = predict_next_word_smoothed(w1, w2, training_trigram_frequencies, training_vocabulary_size)
    if predicted_w3:
        print(f"Bigram: '{w1} {w2}', Actual Next Word: '{w3_actual}', Predicted Next Word: '{predicted_w3}'")



Perplexity on simulated test set: 4461.6468134546785

Qualitative assessment of predictions on test set bigrams:
Bigram: 'gandhis philosophy', Actual Next Word: 'of', Predicted Next Word: 'coloured'
Bigram: 'philosophy of', Actual Next Word: 'nonviolence', Predicted Next Word: 'religion'
Bigram: 'of nonviolence', Actual Next Word: 'shaped', Predicted Next Word: 'coloured'
Bigram: 'nonviolence shaped', Actual Next Word: 'the', Predicted Next Word: 'coloured'
Bigram: 'shaped the', Actual Next Word: 'indian', Predicted Next Word: 'period'
Bigram: 'the indian', Actual Next Word: 'independence', Predicted Next Word: 'coloured'
Bigram: 'indian independence', Actual Next Word: 'movement', Predicted Next Word: 'coloured'
Bigram: 'independence movement', Actual Next Word: 'an', Predicted Next Word: 'coloured'
Bigram: 'movement an', Actual Next Word: 'example', Predicted Next Word: 'coloured'
Bigram: 'an example', Actual Next Word: 'of', Predicted Next Word: 'is'
Bigram: 'example of', Actual Ne

## Summary:

### Data Analysis Key Findings

*   The combined corpus from the selected Wikipedia pages resulted in 4977 unique words.
*   The most common trigrams in the combined corpus include "the study of", "the history of", and "as well as".
*   The calculated perplexity for the trigram model with Add-One smoothing on the simulated test set was high at 4461.65, indicating poor predictive performance on this dataset.
*   Qualitative testing showed that the model frequently predicted the word "coloured" regardless of the preceding bigram.

### Insights or Next Steps

*   The model's high perplexity and the frequent prediction of a single word ("coloured") strongly suggest that the current corpus size is insufficient and significantly impacts the model's ability to learn meaningful patterns.
*   Future work should focus on obtaining a much larger and more diverse text corpus and exploring more sophisticated smoothing techniques (like Kneser-Ney) to address the data sparsity issues inherent in n-gram models.


# Task
Build a trigram model for next word prediction using the `wikipedia` and `gutenberg` libraries for a diverse corpus, applying Kneser-Ney smoothing to address data sparsity issues.

## Install necessary libraries

### Subtask:
Install the necessary libraries using pip.


**Reasoning**:
Attempt to install the `gutenberg` library again, this time in a separate cell, which sometimes helps with dependency resolution in environments like Colab. If it still fails, this will confirm the dependency issue.



In [ ]:
%pip install gutenberg

  Using cached Gutenberg-0.8.2-py3-none-any.whl.metadata (9.4 kB)
  Using cached SPARQLWrapper-2.0.0-py3-none-any.whl.metadata (2.0 kB)
  Using cached bsddb3-6.2.9.tar.gz (230 kB)
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


**Reasoning**:
The installation of `gutenberg` failed again due to the `bsddb3` dependency. I will try to install `bsddb3` separately to see if that resolves the issue or provides a more specific error message for the dependency problem.



In [ ]:
%pip install bsddb3

  Using cached bsddb3-6.2.9.tar.gz (230 kB)
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.


## Install necessary libraries

### Subtask:
Install the necessary libraries using pip.


## Load and preprocess data

### Subtask:
Load text data from a diverse corpus using the `wikipedia` and `gutenberg` libraries, and preprocess it by converting to lowercase, removing punctuation, and tokenizing into words.


**Reasoning**:
Load text data from multiple Wikipedia pages and preprocess it by converting to lowercase, removing punctuation, and tokenizing into words. Since the previous attempts to install the `gutenberg` library failed, I will proceed without it, focusing only on the Wikipedia data as per the instructions, acknowledging the inability to use Gutenberg.



In [ ]:
import wikipedia
import string

# Create a list of Wikipedia page titles for a diverse corpus
page_titles = ["History", "Science", "Art", "Technology", "Literature", "Philosophy", "Music", "Sports", "Politics"]
combined_content = ""

# Iterate through the list of page titles and get content
print("Loading data from Wikipedia...")
for page_title in page_titles:
    print(f"Processing page: {page_title}")
    try:
        page_content = wikipedia.page(page_title).content
        combined_content += page_content
    except wikipedia.exceptions.PageError:
        print(f"Page '{page_title}' not found.")
    except wikipedia.exceptions.DisambiguationError as e:
        print(f"Disambiguation page for '{page_title}'. Options: {e.options}")
        try:
            # Attempt to get content from the first option
            page_content = wikipedia.page(e.options[0]).content
            combined_content += page_content
        except wikipedia.exceptions.PageError:
            print(f"Could not get content from the first option for '{page_title}'.")
    except Exception as e:
        print(f"An unexpected error occurred for page '{page_title}': {e}")

# Convert to lowercase
lowercase_text = combined_content.lower()

# Remove punctuation
translator = str.maketrans('', '', string.punctuation)
cleaned_text = lowercase_text.translate(translator)

# Tokenize into words
words = cleaned_text.split()

print(f"\nSuccessfully loaded and preprocessed data from Wikipedia.")
print(f"Total number of words: {len(words)}")
print(f"First 20 tokens from the diverse corpus: {words[:20]}")

# Since gutenberg installation failed in previous attempts, we will proceed without it.
print("\nNote: The gutenberg library could not be installed in previous steps, so only Wikipedia data is used.")

Loading data from Wikipedia...
Processing page: History
Processing page: Science


/usr/local/lib/python3.12/dist-packages/wikipedia/wikipedia.py:389: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system ("lxml"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.

The code that caused this warning is on line 389 of the file /usr/local/lib/python3.12/dist-packages/wikipedia/wikipedia.py. To get rid of this warning, pass the additional argument 'features="lxml"' to the BeautifulSoup constructor.

  lis = BeautifulSoup(html).find_all('li')


Disambiguation page for 'Science'. Options: ['Sincé', 'Sińce', 'Since (film)', 'Since (album)', 'Since (rapper)', 'Snice']
Processing page: Art
Processing page: Technology
Processing page: Literature
Processing page: Philosophy
Processing page: Music
Page 'Music' not found.
Processing page: Sports
Processing page: Politics
Page 'Politics' not found.

Successfully loaded and preprocessed data from Wikipedia.
Total number of words: 31393
First 20 tokens from the diverse corpus: ['history', 'is', 'the', 'systematic', 'study', 'of', 'the', 'past', 'focusing', 'primarily', 'on', 'the', 'human', 'past', 'as', 'an', 'academic', 'discipline', 'it', 'analyses']

Note: The gutenberg library could not be installed in previous steps, so only Wikipedia data is used.


## Create trigrams

### Subtask:
Generate trigrams (sequences of three consecutive words) from the tokenized text.


**Reasoning**:
Generate trigrams from the tokenized words list and print the first 10 trigrams as per the instructions.



In [ ]:
trigrams = []
for i in range(len(words) - 2):
    trigram = (words[i], words[i+1], words[i+2])
    trigrams.append(trigram)

print("First 10 trigrams:")
for trigram in trigrams[:10]:
    print(trigram)

First 10 trigrams:
('history', 'is', 'the')
('is', 'the', 'systematic')
('the', 'systematic', 'study')
('systematic', 'study', 'of')
('study', 'of', 'the')
('of', 'the', 'past')
('the', 'past', 'focusing')
('past', 'focusing', 'primarily')
('focusing', 'primarily', 'on')
('primarily', 'on', 'the')


## Build trigram frequency table

### Subtask:
Create a frequency distribution of the trigrams to count how many times each trigram appears.


**Reasoning**:
Import Counter and build the trigram frequency table using the list of trigrams, then print the 10 most common trigrams as per the instructions.



In [ ]:
from collections import Counter

trigram_frequencies = Counter(trigrams)

print("10 most common trigrams:")
for trigram, count in trigram_frequencies.most_common(10):
    print(f"{trigram}: {count}")

10 most common trigrams:
('as', 'well', 'as'): 22
('the', 'study', 'of'): 21
('such', 'as', 'the'): 20
('the', 'history', 'of'): 19
('the', 'philosophy', 'of'): 17
('the', 'invention', 'of'): 15
('the', 'use', 'of'): 14
('the', 'development', 'of'): 13
('in', 'the', 'form'): 12
('the', 'form', 'of'): 12


## Build next word prediction function

### Subtask:
Create a function that takes two words as input and predicts the next word based on the trigram frequencies, incorporating Kneser-Ney smoothing.


**Reasoning**:
Define the function to predict the next word using Kneser-Ney smoothing based on trigram frequencies, calculate the vocabulary size, and test the function with example word pairs as per the instructions.



In [ ]:
from collections import Counter
import numpy as np

def predict_next_word_kneser_ney(word1, word2, trigram_frequencies, words):
    """
    Predicts the next word based on trigram frequencies with Kneser-Ney smoothing.

    Args:
        word1: The first word of the bigram.
        word2: The second word of the bigram.
        trigram_frequencies: A Counter object containing trigram frequencies.
        words: The full list of words in the corpus for vocabulary and context counts.

    Returns:
        The most likely next word based on Kneser-Ney smoothing, or None if no
        matching bigrams are found and the vocabulary is empty.
    """
    vocabulary = set(words)
    vocabulary_size = len(vocabulary)

    # Calculate count(w1, w2) - occurrences of the bigram
    bigram_freq = sum(1 for trigram in trigram_frequencies if trigram[0] == word1 and trigram[1] == word2)

    # Calculate count(w1, w2, w3) for all w3 following (w1, w2)
    matching_trigrams = {trigram[2]: freq for trigram, freq in trigram_frequencies.items() if trigram[0] == word1 and trigram[1] == word2}

    # Calculate the number of unique words that follow w2 (for the continuation probability)
    words_following_w2 = set(trigram[2] for trigram in trigram_frequencies if trigram[1] == word2)
    count_unique_following_w2 = len(words_following_w2)

    # Calculate the total number of unique bigrams in the corpus
    unique_bigrams = set((trigram[0], trigram[1]) for trigram in trigram_frequencies)
    total_unique_bigrams = len(unique_bigrams)

    # Discount factor (simplified for this example, can be learned from data)
    discount = 0.75

    smoothed_probabilities = {}

    for w3 in vocabulary:
        # Count(w1, w2, w3)
        trigram_count = trigram_frequencies.get((word1, word2, w3), 0)

        # Calculate the first part of the Kneser-Ney formula (discounted probability)
        if bigram_freq > 0:
            discounted_prob = max(0, trigram_count - discount) / bigram_freq
        else:
            discounted_prob = 0 # Cannot calculate discounted prob if bigram never occurred

        # Calculate the continuation probability P_cont(w3)
        # P_cont(w3) is proportional to the number of unique bigrams ending with w3
        count_bigrams_ending_with_w3 = len(set((trigram[0], trigram[1]) for trigram in trigram_frequencies if trigram[2] == w3))

        if total_unique_bigrams > 0:
             p_cont_w3 = count_bigrams_ending_with_w3 / total_unique_bigrams
        else:
             p_cont_w3 = 0


        # Calculate the interpolation weight (lambda)
        # lambda(w1, w2) = (discount / count(w1, w2)) * |{w3: count(w1, w2, w3) > 0}|
        # simplified: (discount / count(w1, w2)) * number of unique words that follow (w1, w2)
        num_unique_followers = len(matching_trigrams)

        if bigram_freq > 0:
            lambda_w1_w2 = (discount / bigram_freq) * num_unique_followers
        else:
            # If bigram_freq is 0, the lambda term is 1, and the probability is solely based on continuation
             lambda_w1_w2 = 1


        # Combine the two parts using the interpolation weight
        smoothed_prob = discounted_prob + lambda_w1_w2 * p_cont_w3

        smoothed_probabilities[w3] = smoothed_prob

    if smoothed_probabilities:
        # Find the word with the highest smoothed probability
        predicted_word = max(smoothed_probabilities, key=smoothed_probabilities.get)
        return predicted_word
    else:
        return None # Should not happen if vocabulary is not empty

# Calculate the vocabulary size
# vocabulary_size = len(set(words)) # Already calculated inside the function

# Test the predict_next_word_kneser_ney function with some example word pairs
print(f"Vocabulary size: {len(set(words))}")

test_cases = [
    ("the", "study"),
    ("history", "is"),
    ("as", "well"),
    ("art", "history"),
    ("technology", "and"),
    ("nonexistent", "bigram"), # Example of words not expected to be found
    ("of", "the") # Common bigram
]

for word1, word2 in test_cases:
    prediction = predict_next_word_kneser_ney(word1, word2, trigram_frequencies, words)
    print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

Vocabulary size: 5685
Given 'the study', the predicted next word is: of
Given 'history is', the predicted next word is: a
Given 'as well', the predicted next word is: as
Given 'art history', the predicted next word is: and
Given 'technology and', the predicted next word is: a
Given 'nonexistent bigram', the predicted next word is: the
Given 'of the', the predicted next word is: the


## Test the model

### Subtask:
Test the prediction function with some example word pairs.


**Reasoning**:
Call the predict_next_word_kneser_ney function with different word pairs and print the results to test its functionality, fulfilling the subtask requirements.



In [ ]:
# Test the predict_next_word_kneser_ney function with some example word pairs
test_cases = [
    ("the", "study"),
    ("history", "is"),
    ("as", "well"),
    ("art", "history"),
    ("technology", "and"),
    ("nonexistent", "bigram"), # Example of words not expected to be found
    ("of", "the"), # Common bigram
    ("in", "the"), # Another common bigram
    ("science", "and"), # Bigram related to one of the corpus topics
    ("literature", "is") # Bigram related to another corpus topic
]

print("Testing predict_next_word_kneser_ney function:")
for word1, word2 in test_cases:
    prediction = predict_next_word_kneser_ney(word1, word2, trigram_frequencies, words)
    print(f"Given '{word1} {word2}', the predicted next word is: {prediction}")

Testing predict_next_word_kneser_ney function:
Given 'the study', the predicted next word is: of
Given 'history is', the predicted next word is: a
Given 'as well', the predicted next word is: as
Given 'art history', the predicted next word is: and
Given 'technology and', the predicted next word is: a
Given 'nonexistent bigram', the predicted next word is: the
Given 'of the', the predicted next word is: the
Given 'in the', the predicted next word is: form
Given 'science and', the predicted next word is: the
Given 'literature is', the predicted next word is: a


## Evaluate and refine

### Subtask:
Evaluate the model's performance on the diverse corpus and explore ways to further refine the model.


**Reasoning**:
Split the data, re-calculate frequencies on the training set, implement perplexity calculation with Kneser-Ney smoothing, calculate perplexity, and perform qualitative assessment, then discuss findings as per the instructions.



In [ ]:
import numpy as np

# 1. Split the available words data into a training set and a small test set
split_ratio = 0.8
split_index = int(len(words) * split_ratio)
training_words = words[:split_index]
test_words = words[split_index:]

print(f"Training set size: {len(training_words)} words")
print(f"Test set size: {len(test_words)} words")

# 2. Re-calculate the trigram_frequencies and the vocabulary based *only* on the training set.
training_trigrams = []
for i in range(len(training_words) - 2):
    trigram = (training_words[i], training_words[i+1], training_words[i+2])
    training_trigrams.append(trigram)

training_trigram_frequencies = Counter(training_trigrams)
training_vocabulary = set(training_words)
training_vocabulary_size = len(training_vocabulary)

print(f"Training vocabulary size: {training_vocabulary_size}")

# Helper function to calculate Kneser-Ney smoothed probability P(w3 | w1, w2)
def get_kneser_ney_prob(w1, w2, w3, training_trigram_frequencies, training_words):
    """
    Calculates the Kneser-Ney smoothed probability P(w3 | w1, w2).

    Args:
        w1: The first word of the bigram.
        w2: The second word of the bigram.
        w3: The potential next word.
        training_trigram_frequencies: Counter of trigram frequencies from training data.
        training_words: List of words in the training corpus.

    Returns:
        The Kneser-Ney smoothed probability P(w3 | w1, w2).
    """
    training_vocabulary = set(training_words)
    training_vocabulary_size = len(training_vocabulary)
    discount = 0.75 # Using a fixed discount for simplicity

    # Count(w1, w2, w3)
    trigram_count = training_trigram_frequencies.get((w1, w2, w3), 0)

    # Count(w1, w2) - occurrences of the bigram in training data
    # We can approximate this by summing frequencies of all trigrams starting with (w1, w2)
    bigram_freq = sum(freq for (t1, t2, t3), freq in training_trigram_frequencies.items() if t1 == w1 and t2 == w2)

    # Number of unique words that follow (w1, w2) in training data
    num_unique_followers = len(set(t3 for (t1, t2, t3) in training_trigram_frequencies if t1 == w1 and t2 == w2))

    # Calculate the first part of the Kneser-Ney formula (discounted probability)
    if bigram_freq > 0:
        discounted_prob = max(0, trigram_count - discount) / bigram_freq
    else:
        discounted_prob = 0 # Cannot calculate discounted prob if bigram never occurred

    # Calculate the continuation probability P_cont(w3)
    # P_cont(w3) is proportional to the number of unique bigrams ending with w3
    count_bigrams_ending_with_w3 = len(set((t1, t2) for (t1, t2, t3) in training_trigram_frequencies if t3 == w3))

    # Total number of unique bigrams in the training corpus
    unique_bigrams_training = set((t1, t2) for (t1, t2, t3) in training_trigram_frequencies)
    total_unique_bigrams_training = len(unique_bigrams_training)

    if total_unique_bigrams_training > 0:
         p_cont_w3 = count_bigrams_ending_with_w3 / total_unique_bigrams_training
    else:
         p_cont_w3 = 1 / training_vocabulary_size # Fallback if no bigrams exist (unlikely with sufficient data)


    # Calculate the interpolation weight (lambda)
    if bigram_freq > 0:
        lambda_w1_w2 = (discount / bigram_freq) * num_unique_followers
    else:
        # If bigram_freq is 0, the lambda term is 1, and the probability is solely based on continuation
         lambda_w1_w2 = 1

    # Combine the two parts using the interpolation weight
    smoothed_prob = discounted_prob + lambda_w1_w2 * p_cont_w3

    return smoothed_prob if smoothed_prob > 0 else 1e-10 # Return a small positive value for log calculation

# 3. Implement a function to calculate perplexity on the test set
def calculate_perplexity(test_words, training_trigram_frequencies, training_words):
    """
    Calculates perplexity on a test set using Kneser-Ney smoothed probabilities.

    Args:
        test_words: List of words in the test corpus.
        training_trigram_frequencies: Counter of trigram frequencies from training data.
        training_words: List of words in the training corpus.

    Returns:
        The perplexity score, or None if not enough test trigrams.
    """
    log_prob_sum = 0
    n_test_trigrams = 0

    # Iterate through trigrams in the test set
    for i in range(len(test_words) - 2):
        w1, w2, w3 = test_words[i], test_words[i+1], test_words[i+2]

        # Calculate smoothed probability P(w3 | w1, w2) using the training model
        smoothed_prob = get_kneser_ney_prob(w1, w2, w3, training_trigram_frequencies, training_words)

        log_prob_sum += np.log2(smoothed_prob)
        n_test_trigrams += 1

    if n_test_trigrams > 0:
        average_log_prob = log_prob_sum / n_test_trigrams
        perplexity = 2 ** (-average_log_prob)
        return perplexity
    else:
        return None

# 4. Calculate and print the perplexity on the test set.
perplexity = calculate_perplexity(test_words, training_trigram_frequencies, training_words)

if perplexity is not None:
    print(f"\nPerplexity on the test set (Kneser-Ney smoothing): {perplexity}")
else:
    print("\nNot enough test trigrams to calculate perplexity.")

# 5. Qualitatively assess the model's predictions on a few bigrams from the test set
print("\nQualitative assessment of predictions on test set bigrams:")
# Select a few bigrams from the test set to test
test_bigrams_for_assessment = []
for i in range(len(test_words) - 2):
    test_bigrams_for_assessment.append((test_words[i], test_words[i+1], test_words[i+2]))
    if len(test_bigrams_for_assessment) >= 10: # Assess the first 10 bigrams in the test set
        break

# Use the previously defined predict_next_word_kneser_ney function for qualitative assessment
# We need to redefine it to use the training data frequencies and words
def predict_next_word_kneser_ney_assessment(word1, word2, training_trigram_frequencies, training_words):
    """
    Predicts the next word based on trigram frequencies with Kneser-Ney smoothing,
    using the training data.

    Args:
        word1: The first word of the bigram.
        word2: The second word of the bigram.
        training_trigram_frequencies: A Counter object containing trigram frequencies from training data.
        training_words: The list of words in the training corpus for vocabulary and context counts.

    Returns:
        The most likely next word based on Kneser-Ney smoothing, or None if no
        matching bigrams are found and the vocabulary is empty.
    """
    training_vocabulary = set(training_words)

    # Calculate smoothed probabilities for all possible next words
    smoothed_probabilities = {}
    for w3 in training_vocabulary:
        smoothed_probabilities[w3] = get_kneser_ney_prob(word1, word2, w3, training_trigram_frequencies, training_words)

    if smoothed_probabilities:
        # Find the word with the highest smoothed probability
        predicted_word = max(smoothed_probabilities, key=smoothed_probabilities.get)
        return predicted_word
    else:
        return None # Should not happen if vocabulary is not empty


for w1, w2, w3_actual in test_bigrams_for_assessment:
    predicted_w3 = predict_next_word_kneser_ney_assessment(w1, w2, training_trigram_frequencies, training_words)
    print(f"Bigram: '{w1} {w2}', Actual Next Word: '{w3_actual}', Predicted Next Word: '{predicted_w3}'")

# 6. Discuss the model's performance and identify potential areas for refinement
print("\n--- Model Evaluation and Refinement Discussion ---")
print(f"Perplexity on the test set: {perplexity:.2f}" if perplexity is not None else "Perplexity could not be calculated.")
print("\nQualitative Assessment Observations:")
print("- Observe how often the predicted word matches the actual next word.")
print("- Note if the predicted words seem contextually relevant.")
print("- Pay attention to predictions for bigrams that might be rare in the training data but present in the test set.")

print("\nDiscussion and Potential Refinements:")
print("1. Perplexity:")
print(f"   - A perplexity of {perplexity:.2f} (if calculated) indicates how well the model predicts the test data." if perplexity is not None else "   - Perplexity could not be calculated.")
print("   - Lower perplexity is better. The value provides a quantitative measure of the model's uncertainty.")
print("2. Qualitative Assessment:")
print("   - The qualitative assessment provides insights into the types of predictions the model makes.")
print("   - Do the predictions make sense in context? Are they generic or specific?")
print("3. Limitations and Refinements:")
print("   - Corpus Size and Diversity: The current corpus, while more diverse than a single page, is still relatively small.")
print("     - Refinement: Train on a much larger and more varied text corpus (e.g., Project Gutenberg texts if installation issues are resolved, or other large datasets).")
print("   - Kneser-Ney Smoothing: While better than Add-One, Kneser-Ney has a discount factor that can be tuned.")
print("     - Refinement: Implement a method to learn the optimal discount factor from the training data.")
print("   - Higher-Order N-grams: Using 4-grams or 5-grams might capture longer-range dependencies but would require significantly more data.")
print("     - Refinement: Consider higher-order models if a much larger corpus is available.")
print("   - Backoff/Interpolation: Kneser-Ney is a form of interpolation, but more explicit backoff or interpolation with lower-order models (bigrams, unigrams) could be explored.")
print("     - Refinement: Formalize the backoff mechanism to ensure robust predictions even for unseen bigrams.")
print("   - Handling Out-of-Vocabulary (OOV) Words: Words in the test set not seen in training are currently not handled explicitly.")
print("     - Refinement: Implement a strategy for OOV words, such as replacing them with a special <UNK> token during preprocessing or using techniques like character-level models.")
print("   - Context Length: Trigrams only consider the previous two words. For complex language, longer context might be beneficial.")
print("     - Refinement: Explore neural network-based language models (like LSTMs or Transformers) that can capture longer-range dependencies, although this is a significant shift from n-gram models.")

print("\nOverall, the model's performance is likely limited by the corpus size. Increasing the training data is the most crucial next step. Further refinements like tuning smoothing parameters and exploring more advanced techniques can be pursued after obtaining a larger corpus.")

Training set size: 25114 words
Test set size: 6279 words
Training vocabulary size: 4835

Perplexity on the test set (Kneser-Ney smoothing): 12184.195202525192

Qualitative assessment of predictions on test set bigrams:
Bigram: 'whether the', Actual Next Word: 'past', Predicted Next Word: 'source'
Bigram: 'the past', Actual Next Word: 'fully', Predicted Next Word: 'in'
Bigram: 'past fully', Actual Next Word: 'determines', Predicted Next Word: 'the'
Bigram: 'fully determines', Actual Next Word: 'the', Predicted Next Word: 'the'
Bigram: 'determines the', Actual Next Word: 'present', Predicted Next Word: 'the'
Bigram: 'the present', Actual Next Word: 'and', Predicted Next Word: 'and'
Bigram: 'present and', Actual Next Word: 'what', Predicted Next Word: 'in'
Bigram: 'and what', Actual Next Word: 'implications', Predicted Next Word: 'the'
Bigram: 'what implications', Actual Next Word: 'this', Predicted Next Word: 'the'
Bigram: 'implications this', Actual Next Word: 'would', Predicted Next Wo

pip install gutenberg